File: built-in plugins dsp/mb-comp-studio.cpp
/**
 * MB Studio Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Clean studio compressor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_STUDIO_H
#define MB_COMP_STUDIO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompStudio : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-studio";
    static constexpr const char* PLUGIN_NAME    = "MB Studio Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -20f;  // range [-60, 0]
    float ratio = 4f;  // range [1, 20]
    float attack = 10f;  // range [0.1, 100]
    float release = 100f;  // range [10, 1000]
    };

    MbCompStudio() = default;
    ~MbCompStudio() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -60f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 20f);
        params.attack = std::clamp(params.attack, 0.1f, 100f);
        params.release = std::clamp(params.release, 10f, 1000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Studio Comp
        return input;
    }
};

#endif // MB_COMP_STUDIO_H

File: built-in plugins dsp/mb-comp-studio.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCompStudioPlugin: PluginDefinition = { id: 'mb-comp-studio', slug: 'mb-comp-studio', name: 'MB Studio Comp', category: 'effect', type: 'compressor', version: '1.0.0', description: 'Clean studio compressor', author: 'Max Booster', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -20, minValue: -60, maxValue: 0, automatable: true }, { id: 'ratio', name: 'Ratio', type: 'float', defaultValue: 4, minValue: 1, maxValue: 20, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 10, minValue: 0.1, maxValue: 100, automatable: true }, { id: 'release', name: 'Release', type: 'float', defaultValue: 100, minValue: 10, maxValue: 1000, automatable: true }], defaultPreset: { threshold: -20, ratio: 4, attack: 10, release: 100 } };

export default MbCompStudioPlugin;
